src/mpid/ch4/netmod/ofi/ofi_huge.c

```diff
@@ -175,8 +175,20 @@ int MPIDI_OFI_peek_huge_event(int vni, struct fi_cq_tagged_entry *wc, MPIR_Reque
             found_msg = true;
         }
     }
-    if (!found_msg) {
-        /* FIXME: the count is wrong in this case. We need progress until the control message is received */
+    if (found_msg) {
+        rreq->status.MPI_SOURCE = MPIDI_OFI_cqe_get_source(wc, false);
+        rreq->status.MPI_TAG = MPIDI_OFI_init_get_tag(wc->tag);
+        rreq->status.MPI_ERROR = MPI_SUCCESS;
+        MPIR_STATUS_SET_COUNT(rreq->status, count);
+        /* util_id should be the last thing to change in rreq. Reason is
+         * we use util_id to indicate peek_event has completed and all the
+         * relevant values have been copied to rreq. */
+        MPL_atomic_release_store_int(&(MPIDI_OFI_REQUEST(rreq, util_id)), MPIDI_OFI_PEEK_FOUND);
+    } else if (MPIDI_OFI_REQUEST(rreq, kind) == MPIDI_OFI_req_kind__probe) {
+        /* return not found for this probe. User can probe again. */
```

Comment:
`probe` is independent of later `receive`, thus we can simply discard the `probe`. In the case when `Iprobe` sees a message but missing its "control header", it can pretend it didn't see the message and most likely there will be a next `Iprobe` anyway.

`mprobe` is trickier but I may be confused about its semantics --
* Once `mprobe` succeeds, does a later regular `recv` with the same envelope get the same message? My impression was no, but let's clarify it.
* My interpretation was a successful `mprobe` sort of claimed the message than only a later `mrecv` will get that message. Since it is being claimed, a second `mprobe` shouldn't get the same message.
* You seem to indicate that user can issue multiple `mprobe` and obtain the same message handle? Are the allowed to issue multiple `mrecv` on that same message handle then?

Reply:
Quote from standard --
```
... except that MPI_MPROBE and MPI_IMPROBE provide a mechanism to
receive the specific message that was matched regardless of other intervening probe or
receive operations.
```
That, to me, indicates that another `mprobe` is not allowed to getting the same message as another successful `mprobe` -- where "intervening probe" I take it includes "intervening mprobe".

Reply:
> * You seem to indicate that the user can issue multiple `mprobe` and obtain the same message handle? Are they allowed to issue multiple `mrecv` on that same message handle then?

What I'm saying is that if the control header is missing, just return message not found. If its a blocking mprobe, then the upper layer will call back down after poking progress.

Reply:
With `mprobe` we call libfabric with `FI_CLAIM` flag, which I believe will prevent the next `mprobe` (from the upper layer) to get the message. Libfabric is not aware of control header, so if we discard a matching `mprobe`, that message is essentially lost.

Reply:
That's what I was missing. The probe/claim is at the libfabric level. I think I got it now.